proposals: Add Volume Hostpath "type" proposal

euank opened the pull request:
This is a continuation of #31384. It's related to #26816 as well.
The discussion in #31384 is worth reading and this proposal largely derives from my comments there.

cc @thockin @pmorie @saad-ali @kubernetes/sig-storage 
cc @yujuhong since it talks briefly about kubelet doing more

cc @calebamiles I think we might need a "Feature" for this since it's an api change, though a minor one?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34058)

<!-- Reviewable:end -->

euank committed:
proposals: Add Volume Hostpath "type" proposal

smarterclayton commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,96 @@
+| `exists` | If nothing exists at the given path, the pod will fail to run and provide an informative error message |
+| `file` | If a file does not exist at the given path, the pod will fail to run and provide an informative error message |
+| `device` | If a block or character device does not exist at the given path, the pod will fail to run and provide an informative error message |
+| `socket` | If a socket does not exist at the given path, the pod will fail to run and provide an informative error message |
```
If I want to share a socket between two containers in the same pod, would I want the same support here for that?  Or are we saying emptyDir and intra-pod are less challenged than pod to host?

euank replied:
Within a pod, it's the responsibility of those two processes to synchronize on the socket themselves.

The root problem, of some random directory being created if the socket doesn't exist, doesn't happen in that instance because the containers presumably don't implement docker's logic of "exists || mkdir", and it's well understood that the emptyDir exists as a directory with nothing in it.

intrapod is also less challenging because it isn't implicitly relying on some external state/process existing.

smarterclayton commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,96 @@
+While I propose exposing the same set of possible values for consistency, in reality a `SubPath` is useful primarily under a network mount. Because of this, the `device` and `socket` values would likely not be useful.
+
+I do not have any examples of concrete examples from the Kubernetes project for this, in part because there are very few examples of `SubPath` to begin with.
+However, each of the use-cases described for Host Volumes also could conceivably occur as subdirectories of a network mount using this feature, and thus still offer insight into why it would be of value.
```
I'm aware of folks using it to share host NFS mounts in predictable ways across a cluster (I think)

smarterclayton replied:
Will double check.

smarterclayton commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,96 @@
+
+### Containerized Kubelet
+
+A containerized kubelet would have difficulty creating directories. The implementation will likely respect the `containerized` flag, or similar, allowing it to either break out or be "/rootfs/" aware and thus operate as desired.
```
Containerized kubelets effectively are required to have access to host rootfs in practice (that's the only variant we support in production at Red Hat), so seems ok.

foxish commented:
adding issue comment.

euank commented:
I added an alternative possible API that reduces it to adding a single `MustExist` bool. I think the gain in simplicity is likely worth the loss in specificity, but a second opinion would be appreciated.

@jonboulle, @pmorie, @thockin if any of you have a strong opinion on the above or cycles to review, it would be appreciated!

jonboulle commented:
Overall, this LGTM!
One question: what permissions and ownership should/would Kubelet apply if it creates a directory?

euank commented:
@jonboulle `root:root` 755 seems like the sane choice. We can narrow them down in the future. There's some discussion related to this [here](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/volumes.md#shared-storage-hostpath) and this change doesn't impact that assessment.

I'll edit it in

thockin commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,185 @@
+conceivably occur as subdirectories of a network mount using this feature, and
+thus still offer insight into why it would be of value.
+
+<!-- uncomment if this comes up :) An argument against providing this feature for SubPaths is that it might not increase correctness as much; while adding such qualifications to Host volumes allows a pod to express a dependency that the node may not meet, adding it to subPath allows the pod to express an assumption about a given filesystem it's mounting. This assumption is not node specific and,  ...
```
I'm inclined to agree - when in doubt, leave it out,

thockin commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,185 @@
+
+| Value | Behavior |
+|:------|:---------|
+| `auto ` | If nothing exists at the given path, an empty directory will be created there. Otherwise, behaves like `exists` |
```
You need to handle existing objects which will spontaneously grow this field, with a default value of "".  Perhaps  "" is how you spell "anything".

Given that, the field might be better named as "require" something, where "" naturally means "nothing" or "no requirements"

euank replied:
I'm not sure I understand. Regarding handling existing objects, above I wrote "If not set, defaults to `auto`" which was meant to cover that. `auto` is the existing behaviour, so it's backwards compatible for existing objects.
Is having `nil == "" == "auto"` a no-no for k8s api?

`require` seems clever for the one case of default, but I think I like `type` more overall. (`hostPath.type = file` vs `hostPath.require = file` seems more natural as `type`).

Mind clarifying what you meant?

thockin reviewed (commented):
LGTM overall

thockin commented:
What I meant: "If not set, defaults to auto" means that any object that
comes through the API will have this field set to "auto".  Objects that
already exist do not get the default-setting hook called, so they exist
with a literal value of "". But upon reflection, I am wrong.  When they are
read from etcd, they get defaults, which handles this case.

I still think it might be cleaner to have "" be the default value, rather
than a literal "auto" - that way it will simply be omitted when the user
doesn't care.

On Thu, Oct 13, 2016 at 10:37 AM, Euan Kemp <notifications@github.com>
wrote:

> *@euank* commented on this pull request.
> ------------------------------
>
> In docs/proposals/volume-hostpath-qualifiers.md
> <https://github.com/kubernetes/kubernetes/pull/34058>:
>
> > +of these volumes into the Kubelet to the benefit of the rkt container runtime.
> +
> +
> +## Proposed API Change
> +
> +### Host Volume
> +
> +I propose that the
> +[`v1.HostPathVolumeSource`](https://github.com/kubernetes/kubernetes/blob/d26b4ca2859aa667ad520fb9518e0db67b74216a/pkg/api/types.go#L447-L451)
> +object be changed to include the following additional field:
> +
> +`Type` - An optional string of `auto|exists|file|device|socket|directory` - If not set, defaults to `auto`. These values have the following behavior:
> +
> +| Value | Behavior |
> +|:------|:---------|
> +| `auto ` | If nothing exists at the given path, an empty directory will be created there. Otherwise, behaves like `exists` |
>
> I'm not sure I understand. Regarding handling existing objects, above I
> wrote "If not set, defaults to auto" which was meant to cover that. auto
> is the existing behaviour, so it's backwards compatible for existing
> objects.
> Is having nil == "" == "auto" a no-no for k8s api?
>
> require seems clever for the one case of default, but I think I like type
> more overall. (hostPath.type = file vs hostPath.require = file seems more
> natural as type).
>
> Mind clarifying what you meant?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34058>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVG40zd5K0Eycvlx3ZSn6F4pG2Zgpks5qzmx2gaJpZM4KOOl9>
> .
>

euank commented:
Updated to remove subpath and change the default to `""` per @thockin's comments. Agreed that that's cleaner, and we can add subpathtype later if people want it.

I also added some words about selinux (cc @pmorie) and permissions at the bottom.

pmorie commented:
This generally LGTM, but I do have one concern that we need to check -- what the SELinux context of artifacts on the host's FS is when containerized Kubelet creates them.

euank commented:
Just squashed, no actual changes..

euank commented:
*sigh* missed a munger lint, no actual changes.

prabhakar-pal commented:
Could you please inform if the hostPath type attribute has been added to kubernetes? If so, in which version?

euank commented:
@prabhakar-pal it has not been implemented yet.

thockin commented:
Damn, I assumed that since we merged this, we must have gotten around to implementing, but I guess not.

This would be a GREAT contribution.

bboreham commented on docs/proposals/volume-hostpath-qualifiers.md:
```diff
@@ -0,0 +1,179 @@
+
+|Value | Behavior | Reason for exclusion |
+|:-----|:---------|:---------------------|
+| `new-directory` | Like `auto`, but the given path must be a directory if it exists | `auto` mostly fills this use-case |
```
`auto` is not defined in this document
